File: python/DVVector.py
from . import PyThrustRTC as native
import numpy as np
from .DeviceViewable import DeviceViewable

class DVVector(DeviceViewable):
	def __init__(self, cptr):
		self.m_cptr = cptr

	def name_elem_cls(self):
		return native.n_dvvector_name_elem_cls(self.m_cptr)

	def size(self):
		return native.n_dvvector_size(self.m_cptr)

	def to_host(self):
		elem_cls = self.name_elem_cls()
		if elem_cls=='int8_t':
			nptype = np.int8
		elif elem_cls=='uint8_t':
			nptype = np.uint8
		elif elem_cls=='int16_t':
			nptype = np.int16
		elif elem_cls=='uint16_t':
			nptype = np.uint16
		elif elem_cls=='int32_t':
			nptype = np.int32
		elif elem_cls=='uint32_t':
			nptype = np.uint32
		elif elem_cls=='int64_t':
			nptype = np.int64
		elif elem_cls=='uint64_t':
			nptype = np.uint64
		elif elem_cls=='float':
			nptype = np.float32
		elif elem_cls=='double':
			nptype = np.float64
		elif elem_cls=='bool':
			nptype = np.bool
		ret = np.empty(self.size(), dtype=nptype)
		native.n_dvvector_to_host(self.m_cptr, ret.__array_interface__['data'][0])
		return ret

def device_vector(ctx, elem_cls, size, ptr_host_data=None):
	return DVVector(native.n_dvvector_create(ctx.m_cptr, elem_cls, size, ptr_host_data))

def device_vector_from_numpy(ctx, nparr):
	if nparr.dtype == np.int8:
		elem_cls = 'int8_t'
	elif nparr.dtype == np.uint8:
		elem_cls = 'uint8_t'
	elif nparr.dtype == np.int16:
		elem_cls = 'int16_t'
	elif nparr.dtype == np.uint16:
		elem_cls = 'uint16_t'
	elif nparr.dtype == np.int32:
		elem_cls = 'int32_t'
	elif nparr.dtype == np.uint32:
		elem_cls = 'uint32_t'		
	elif nparr.dtype == np.int64:
		elem_cls = 'int64_t'
	elif nparr.dtype == np.uint64:
		elem_cls = 'uint64_t'	
	elif nparr.dtype == np.int64:
		elem_cls = 'int64_t'
	elif nparr.dtype == np.uint64:
		elem_cls = 'uint64_t'	
	elif nparr.dtype == np.float32:
		elem_cls = 'float'
	elif nparr.dtype == np.float64:
		elem_cls = 'double'
	elif nparr.dtype == np.bool:
		elem_cls = 'bool'
	size = len(nparr)
	ptr_host_data = nparr.__array_interface__['data'][0]
	return device_vector(ctx, elem_cls, size, ptr_host_data)

def device_vector_from_dvs(ctx, lst_dv):
	return DVVector(native.n_dvvector_from_dvs(ctx.m_cptr, [item.m_cptr for item in lst_dv]))

def device_vector_from_list(ctx, lst, elem_cls):
	devive_viewables = [ native.n_dv_create_basic(elem_cls, item) for item in lst]
	dvec = DVVector(native.n_dvvector_from_dvs(ctx.m_cptr, devive_viewables))
	for dv in devive_viewables:
		native.n_dv_destroy(dv)
	return dvec

File: java/DVVector.java
package JThrustRTC;

public class DVVector extends DVVectorLike
{
	public DVVector(String elem_cls, int size)
	{
		super( Native.dvvector_create(elem_cls, size) );
	}

	public DVVector(byte[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public DVVector(short[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public DVVector(int[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public DVVector(long[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public DVVector(float[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public DVVector(double[] hdata)
	{
		super( Native.dvvector_create(hdata) );
	}

	public Object to_host(int begin, int end)
	{
		String type = name_elem_cls();
		if (end == -1) end = size();
        if (end < begin) return null;
        int h_size = end - begin;

        if (type.equals("int8_t"))
        {
        	byte[] arr = new byte[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }
        else if (type.equals("int16_t"))
        {
        	short[] arr = new short[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }
        else if (type.equals("int32_t"))
        {
        	int[] arr = new int[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }
        else if (type.equals("int64_t"))
        {
        	long[] arr = new long[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }
        else if (type.equals("float"))
        {
        	float[] arr = new float[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }
        else if (type.equals("double"))
        {
        	double[] arr = new double[h_size];
        	if (h_size>0)
            {
            	Native.dvvector_to_host(cptr(), arr, begin, end);
                return arr;
            }
        }

        return null;
	}

	public Object to_host()
	{
		return to_host(0, -1);
	}


}